File: FormalBench-data/FilteredData/specifications/seed_726_per_category_10_653ade686f/c/CountOddSquares.c
#include "java_arrays.h"
#include <math.h>
#include <stdint.h>

/*@
assigns \nothing;
ensures \result == (int32_t)((integer)left + right);
*/
static int32_t java_add(int32_t left, int32_t right) {
    return (int32_t)((uint32_t)left + (uint32_t)right);
}

/*@
logic integer fb_floor_root(integer x, integer lo, integer hi) =
  lo > hi ? hi :
  ((lo + hi) / 2) * ((lo + hi) / 2) <= x ? fb_floor_root(x, (lo + hi) / 2 + 1, hi) : fb_floor_root(x, lo, (lo + hi) / 2 - 1);
logic integer fb_squares_through(integer x) = x < 0 ? 0 : fb_floor_root(x, 0, 46340) + 1;
*/
/*@
terminates m < INT32_MAX;
assigns errno;
ensures m < INT32_MAX;
ensures \result == (n > m ? 0 : fb_squares_through(m) - fb_squares_through((integer)n - 1));
ensures 0 <= \result <= 46341;
ensures n < 0 && n <= m ==> errno == EDOM;
ensures n >= 0 || n > m ==> errno == \old(errno);
*/
int32_t countOddSquares(int32_t n, int32_t m) {
    int32_t count = 0;
    /*@
    loop invariant m < INT32_MAX ==> n <= i <= (n > m ? n : (integer)m + 1);
    loop invariant m < INT32_MAX ==> count == fb_squares_through((integer)i - 1) - fb_squares_through((integer)n - 1);
    loop assigns i, count, errno;
    */
    for (int32_t i = n; i <= m; i = java_add(i, 1)) {
        double square = sqrt((double)i);
        if (fmod(square, 1.0) == 0.0 && i % 1 == 0) {
            count = java_add(count, 1);
        }
    }
    return count;
}

File: FormalBench-data/FilteredData/specifications/seed_726_per_category_10_653ade686f/java/CountOddSquares.java
import java.io.*;
import java.lang.*;
import java.util.*;
import java.math.*;


class CountOddSquares {
    
    /*@
    public normal_behavior
      ensures x < 0 ==> \result == 0;
      ensures x >= 0 ==> 1 <= \result <= 46341
        && (\bigint)(\result - 1) * (\result - 1) <= x
        && x < (\bigint)\result * \result;
    model public static pure int fb_squaresThrough(int x);

    public normal_behavior
      assignable \nothing;
      diverges m == Integer.MAX_VALUE;
      ensures m < Integer.MAX_VALUE;
      ensures \result == (n > m ? 0 : fb_squaresThrough(m) - (n <= 0 ? 0 : fb_squaresThrough(n - 1)));
      ensures 0 <= \result <= 46341;
    @*/
    //@ code_java_math
    public static int countOddSquares(int n, int m) {
        int count = 0;
        /*@
        loop_invariant m < Integer.MAX_VALUE ==> n <= i && (\bigint)i <= (n > m ? (\bigint)n : (\bigint)m + 1);
        loop_invariant m < Integer.MAX_VALUE ==> count == (i == n ? 0 : fb_squaresThrough(i - 1) - (n <= 0 ? 0 : fb_squaresThrough(n - 1)));
        loop_writes i, count;
        @*/
        for (int i = n; i <= m; i++) {
            double square = Math.sqrt(i);
            if (square % 1 == 0 && i % 1 == 0) {
                count++;
            }
        }
        return count;
    }
}

